Main.hs

```diff
@@ -4,30 +4,117 @@ import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = Number Double
+  | Sqrt Expr
+  | Plus Expr Expr
+  | Minus Expr Expr
+  | Prod Expr Expr
+  | Divide Expr Expr
+  | Exp Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Number x) = show x
+  show (Sqrt x) = "sqrt(" ++ show x ++ ")"
+  show (Plus a b) = "(" ++ show a ++ ") + (" ++ show b ++ ")"
+  show (Minus a b) = "(" ++ show a ++ ") - (" ++ show b ++ ")"
+  show (Prod a b) = "(" ++ show a ++ ") * (" ++ show b ++ ")"
+  show (Divide a b) = "(" ++ show a ++ ") / (" ++ show b ++ ")"
+  show (Exp a b) = "(" ++ show a ++ ") ^ (" ++ show b ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (==) :: Expr -> Expr -> Bool
+  (==) a b = show a == show b
 
 data Error
+  = DivisionByZero Expr
+  | RootOfNegative Expr
 
 instance Show Error where
-  show = undefined
+  show e = "ERROR! " ++ case e of
+    DivisionByZero expr -> "Cannot divide by zero in \"" ++ show expr ++ "\""
+    RootOfNegative expr -> "Cannot take a root of a negative number in \"" ++ show expr ++ "\""
 
 instance Eq Error where
-  (==) = undefined
+  (==) (DivisionByZero exp1) (DivisionByZero exp2) = exp1 == exp2
+  (==) (RootOfNegative exp1) (RootOfNegative exp2) = exp1 == exp2
+  (==) _ _ = False
+
+evalBinaryOperands :: Expr -> Expr -> Either Error (Double, Double)
+evalBinaryOperands a b =
+  case eval a of
+    Left err -> Left err
+    Right eval_a ->
+      case eval b of
+        Left err -> Left err
+        Right eval_b -> Right (eval_a, eval_b)
```

Comment:
Good first step to abstracting away binary operations. I would also parameterize this with the function to apply to the arguments when they are successfully computed.